### PR TITLE
Specify /admin/whois/{userId}

### PR DESCRIPTION
**Pull request by illicitonion**

### CODE REVIEW

**Comment by dbkr** [comment]
Was the intention of this API that it became part of the Matrix spec or was it only ever intended to be a synapse-specific admin API? I guess the fact that's it's under the client-server path would imply the former.

**Comment by illicitonion** [comment]
I asked Matthew yesterday whether it should be in the spec and he said yes.

Got a case for no?

File: specification/modules/admin.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Server administration
```
Comment (richvdh) [inline specification/modules/admin.rst]: need to add this to  specification/feature_profiles.rst

File: api/client-server/admin.yaml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            type: object
+            properties:
+              user_id:
+                type: string
+                description: The Matrix user ID of the user.
+              devices:
+                type: array
+                items:
+                  type: object
+                  title: Information about a particular device.
```
Comment (richvdh) [inline api/client-server/admin.yaml]: the 'title' gets used as a 'type' identifier in the generated doc, so it looks somewhat anomalous as a big long name. Here `DeviceInfo` would be more appropriate. Similarly for the other titles

File: api/client-server/admin.yaml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                  properties:
+                    device_id:
```
Comment (richvdh) [inline api/client-server/admin.yaml]: this isn't present in the example. which is lying?

File: api/client-server/admin.yaml (hunk 2 above)
Comment (richvdh) [inline api/client-server/admin.yaml]: It would be nice to give it a description too.

File: api/client-server/admin.yaml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                      type: string
+                      description: |-
+                        An identifier for a device used by the user.
+                        Note that this is always null, because devices are not currently specified.
```
Comment (richvdh) [inline api/client-server/admin.yaml]: is it really null? or is it absent? or empty?

File: api/client-server/admin.yaml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                        ]
+                      }
+                    ]
+                  }
+                ]
+              }
+          schema:
+            type: object
+            properties:
+              user_id:
```
Comment (richvdh) [inline api/client-server/admin.yaml]: this isn't present in the example. which is lying?

File: api/client-server/admin.yaml (hunk 4 above)
Comment (illicitonion) [inline api/client-server/admin.yaml]: Currently, it is always null.

Want to change it?

File: api/client-server/admin.yaml (hunk 5 above)
Comment (illicitonion) [inline api/client-server/admin.yaml]: The example is lying.

Done.

File: api/client-server/admin.yaml (hunk 2 above)
Comment (illicitonion) [inline api/client-server/admin.yaml]: Done

File: specification/modules/admin.rst (hunk 1 above)
Comment (illicitonion) [inline specification/modules/admin.rst]: Done

File: api/client-server/admin.yaml (hunk 4 above)
Comment (richvdh) [inline api/client-server/admin.yaml]: No, if it really is null, that's fine. Can you add it to the example, though?

Comment (illicitonion) [inline api/client-server/admin.yaml]: Unfortunately not, the validator complains that null is not of type string, and Swagger doesn't have a concept of nullability.

Which is awful.

Perhaps I should just make it the empty string...

Comment (illicitonion) [inline api/client-server/admin.yaml]: Fixed by making it the empty string key of a dict

**Comment by illicitonion** [comment]
I'm not going to mention that it happens to be empty, seems unnecessarily constraining

**Pull request merged**